new_dosi_adaptation.nlogo

```diff
@@ -452,8 +454,91 @@ to estimate-demand ;run by the observer, asks consumer good firms to estimate de
+    set inputs-received? true
+  ]
+  while [any? turtles with [done-producing? = false]][
+    ask turtles with [inputs-received? = true and done-producing? = false][
+      produce
+    ]
+  ]
+end
+
+to produce ;will be run by a turtle
+  hire-labor
+  let production-quantity (min (map [i -> production-potential i] (table:keys input-information)))
+  let reduction-factor 0
+  let input-type firm-type
+  ifelse member? self consumer-good-firms[
+    set inventories (inventories + production-quantity)
+  ][
+    if quantity-ordered > 0[
+      set reduction-factor (production-quantity / quantity-ordered)
+    ]
+  ]
+  use-inputs production-quantity
+  let total-revenue 0
+  let input-firm-type firm-type
+  ask my-out-links with [purchase-order > 0][
+    let delivery-quantity (purchase-order * reduction-factor)
+    set purchase-order 0
+    let cost-of-order (delivery-quantity * (quantitative-value * (get-index index)))
+    set total-revenue (total-revenue + cost-of-order)
+    ask other-end[
+      set liquid-assets (liquid-assets - cost-of-order)
+      if delivery-quantity > 0[
+        receive-order delivery-quantity input-firm-type cost-of-order
+      ]
+      if not(any? my-in-links with [purchase-order > 0])[
+        set inputs-received? true
+      ]
+    ]
+  ]
+  set liquid-assets (liquid-assets + total-revenue)
+  set done-producing? true
+end
+
+to receive-order [delivery-quantity input-firm-type cost-of-order] ;will be run by a turtle
+  let current-input-stock get-metric "Current-stock" input-firm-type
+  set-metric "Current-stock" input-firm-type (current-input-stock + delivery-quantity)
+  ifelse current-input-stock = 0[
+    set-metric "Average unit cost" input-firm-type (cost-of-order / delivery-quantity)
+  ][
+    let current-average-cost (get-metric "Average unit cost" input-firm-type)
+        let new-average ((current-average-cost * current-input-stock) + (cost-of-order)) / (current-input-stock + delivery-quantity)
+    set-metric "Average unit cost" input-firm-type new-average
+  ]
+end
+
+to use-inputs [production-quantity] ;will be run by a turtle - adjusts all input stocks except labor to reflect production
+  foreach table:keys input-information [i ->
+    if i != 0[
+      let current-stock get-metric "Current-stock" i
+      let mp get-metric "Marginal-productivity" i
+      set-metric "Current-stock" i (current-stock - (production-quantity / mp))
+    ]
+  ]
+end
+
+to hire-labor
+  let labor-quantity-ordered (get-metric "Pending-orders" 0)
+  ifelse TOTAL-LABOR-ORDERS < LABOR-FORCE-SIZE[
+    set-metric "Current-stock" 0 labor-quantity-ordered
+    set liquid-assets (liquid-assets - (labor-quantity-ordered * WAGE-RATE))
+    set LABOR-MONEY-POT (LABOR-MONEY-POT + (labor-quantity-ordered * WAGE-RATE))
+  ][
+    let reduction-factor LABOR-FORCE-SIZE / TOTAL-LABOR-ORDERS
+    set-metric "Current-stock" 0 (reduction-factor * labor-quantity-ordered)
```

Comment:
you calculate (reduction-factor * labor-quantity-ordered) three times in three lines here.